hyperspy/signal.py

```
@@ -3092,7 +3092,12 @@ def get_dimensions_from_data(self):
         """
         dc = self.data
         for axis in self.axes_manager._axes:
-            axis.size = int(dc.shape[axis.index_in_array])
+            new_size = int(dc.shape[axis.index_in_array])
+            if axis.size != new_size:  # Do nothing if already set
+                if not axis.is_uniform:  # Axis must be uniform to set the size
+                    axis.convert_to_uniform_axis()
```

I don't know if it would it be better to raise an error and say that this function doesn't support non-uniform axis? Maybe it would be good to revisit the purpose of this method - I find the name misleading...

---

I'm not a huge fan of this method either but I think that currently it is necessary if you mapped a function on a non-uniform-axis or crop something.  Previously it was kind of a nightmare with Non-Uniform axes as it would return an axis with a different size than the len of the axis.

It is usually called after the ~~`map`~~ `_deepcopy_with_new_data`  function to handle functions which change the size of the axis or when something is stacked/ data is changed.  In general the handling of matching data shape/dimensions and axis shape/dimensions should be handled better.

This might be a larger discussion than just this PR and something that I was meaning to bring up.

The core discussion is how does hyperspy handle the setting of the data attribute in relation to the axes.  Is the axes_manager immediately updated, sizes changed and axes dropped? How does it handle non uniform axes? Are they sliced, repeated, etc. or are they just changed into UniformAxes with offset 0 and scale 1. 

Maybe an `update_axes` function is the best way to go. With the option of passing a list of axes dictionaries might be the best idea and then just have each function call `update_axes` in each situation.

---

Part of the issue is the repeated use of this line of code `self._deepcopy_with_new_data(None)`

This state initializes a signal with no data but retains the axes.

A better behavior might be to formalize how we use the `_deepcopy_with_new_data` function. Specifically how axes are handled. I can make an issue to track this.

---

> This might be a larger discussion than just this PR and something that I was meaning to bring up.

Yes, agreed!

---

see #3041